feat(WebexLocalMedia): replace self-view with local share

alinasuciu opened the pull request:
When the user is sharing the screen the self-view is replaced by the screen share

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-224481

cipak commented on src/components/WebexLocalMedia/WebexLocalMedia.jsx:
```diff
@@ -14,16 +14,18 @@ import {
 /**
- * Webex Local Media component displays the user's local video.
+ * Webex Local Media component displays the user's local video or local share.
  *
```
We should also add an explanatory text of the display logic, eg:

```
If the meeting is active and the current user is sharing the screen, the local share is displayed; otherwise, the local 
video is displayed (if video is enabled).
```

lalli-flores commented on src/components/WebexLocalMedia/WebexLocalMedia.jsx:
```diff
@@ -14,16 +14,20 @@ import {
   const [mediaRef, {width}] = useElementDimensions();
-  const {localVideo} = useMeeting(meetingID);
+  const {localVideo, localShare, remoteVideo} = useMeeting(meetingID);
+  const isActive = remoteVideo !== null;
```
@cipak, @alinasuciu don't forget to add a subtask to update this when refactoring meeting state

alinasuciu replied:
This is not needed anymore as we moved the logic in InMeeting component.

lalli-flores commented on src/components/WebexLocalMedia/WebexLocalMedia.jsx:
```diff
@@ -14,16 +14,20 @@ import {
+  const isActive = remoteVideo !== null;
   const {ID} = useMe();
-  const videoRef = useStream(localVideo);
+  const localStream = (isActive && localShare) || localVideo;
```
I think a more scalable solution for this is to move the logic of deciding what type of local media to show to [In-meeting component](https://github.com/webex/components/blob/master/src/components/WebexInMeeting/WebexInMeeting.jsx), as this component decides what the experience should be when you're in a meeting.
I'm thinking that maybe we can add a prop to local media `mediaType` where an user can pass `video`, `audio` or `screen` and that's how we can determine the type of local media to show

cipak replied:
Initially, we wanted to have two separate components `LocalVideo` and `LocalShare` so that `InMeeting` could choose one of the 2. However, we then saw that `RemoteMedia` displays both share and video, and has some logic of which to display where. So we updated this `LocalMedia` component in a similar way. I still think that, for now, it would be ok to let this component take care of the logic and maybe move it up when the need arises.

lalli-flores replied:
You're right in that `WebexRemoteMedia` has some meeting logic in it, that looking back, shouldn't be there, it should be in `WebexInMeeting` instead. I don't want to make the same mistake with local media, as I foresee some near-team use cases where users just want to get a hold of media display only. Putting the logic there now would basically diminish the usefulness of having it as a component.

lalli-flores commented on src/components/WebexLocalMedia/README.md:
```diff
@@ -1,38 +1,38 @@
+## Preview
+
+To see all the different possible states of the Webex Local Media component,
+you can view our [storybook](https://webex.github.io/components/storybook/?path=/story/webex-local-video--enabled)
```
```suggestion
you can view our [storybook](https://webex.github.io/components/storybook/?path=/story/meetings-webex-local-media--video)

```

lalli-flores commented on src/components/WebexLocalMedia/README.md:
```diff
@@ -1,38 +1,38 @@
+
+2. Create a component instance by passing the meeting ID as a string and enclose it within
+[a data provider](../WebexDataProvider/WebexDataProvider.js) that takes
+the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously
```
```suggestion
the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously. Choose the type of media to display by passing a string to `mediaType`

```

lalli-flores commented on src/components/WebexLocalMedia/WebexLocalMedia.jsx:
```diff
@@ -14,38 +14,54 @@ import {
 WebexLocalMedia.propTypes = {
   className: PropTypes.string,
   meetingID: PropTypes.string.isRequired,
+  mediaType: PropTypes.string,
```
```suggestion
  mediaType: PropTypes. oneOf(['video', 'screen']),
```

lalli-flores commented on src/components/WebexLocalMedia/WebexLocalMedia.stories.js:
```diff
@@ -8,14 +8,22 @@ export default {
 
-export const MediaEnabled = Template.bind({});
-MediaEnabled.args = {
+export const VideoEnabled = Template.bind({});
```
```suggestion
export const Video = Template.bind({});
```

lalli-flores commented on src/components/WebexLocalMedia/WebexLocalMedia.stories.js:
```diff
@@ -8,14 +8,22 @@ export default {
+  mediaType: 'video',
 };
 
 export const MediaDisabled = Template.bind({});
```
```suggestion
export const NoMedia = Template.bind({});
```

lalli-flores commented on src/components/WebexLocalMedia/WebexLocalMedia.stories.js:
```diff
@@ -8,14 +8,22 @@ export default {
+  mediaType: 'video',
+};
+
+export const ShareEnabled = Template.bind({});
```
```suggestion
export const ScreenShare = Template.bind({});
```

lalli-flores reviewed (changes requested):
Few things to address:
* Make sure that the banner and the local video don't overlap. Local video should come under the banner
* Make sure things look proportional on mobile screens
* Arrange the Storybook stories in this order: video, screen share & no media
* Create one commit where you do the local media refactoring / adding the new prop and then another one to do the self-view changes

lalli-flores commented on src/components/WebexInMeeting/WebexInMeeting.scss:
```diff
@@ -16,6 +16,12 @@ $shadow-color: rgba(0, 0, 0, 0.2);
 
+  &.local-sharing {
+    .local-media-in-meeting {
+      top: 2.8rem;
```
How does the transition look when switching between local view and share preview?

lalli-flores reviewed (commented):
@alinasuciu, code looks good. My only concern now is the style override for local sharing preview in In-meeting component. How does the transition look from self-view to preview? does it jump? Maybe all the previews should be aligned at the same level from the top

alinasuciu commented:
> @alinasuciu, code looks good. My only concern now is the style override for local sharing preview in In-meeting component. How does the transition look from self-view to preview? does it jump? Maybe all the previews should be aligned at the same level from the top

I added the same top distance not just when the local share is activated. As the banner has the same height on all devices I had to change the distance from the top for all devices to be the same.

lalli-flores commented:
> I added the same top distance not just when the local share is activated. As the banner has the same height on all devices I had to change the distance from the top for all devices to be the same.

Can you share changes on what that would look like? If the screen is small, it might look weird that the local media is pushed so far down. Is there a way to scale the banner?

Please don't forget to update the Storyshots too since the class `local-sharing` should exist anymore

alinasuciu commented:
> > I added the same top distance not just when the local share is activated. As the banner has the same height on all devices I had to change the distance from the top for all devices to be the same.
> 
> Can you share changes on what that would look like? If the screen is small, it might look weird that the local media is pushed so far down. Is there a way to scale the banner?
> 
> Please don't forget to update the Storyshots too since the class `local-sharing` should exist anymore

@lalli-flores , I updated the stories, I actually forgot to remove the addition of the class if the local share is active, everything should be ok now. I don't see a way to scale the banner, the font is not too big, neither the padding. Doesn't look bad, I would say but it's up to you. 
![Screenshot 2021-04-21 140644](https://user-images.githubusercontent.com/12212457/115545351-60615c00-a2ac-11eb-96dd-47c63c774a2a.png)
![Screenshot 2021-04-21 140728](https://user-images.githubusercontent.com/12212457/115545364-63f4e300-a2ac-11eb-8335-ac85a0016ca8.png)

lalli-flores commented:
@alinasuciu code looks good. One test is failing so once it's fixed this can be merged.

webex-components-publisher commented:
:tada: This PR is included in version 1.61.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@webex/components/v/1.61.0)
- [GitHub release](https://github.com/webex/components/releases/tag/v1.61.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
